File: sip_parser.py
import re
from typing import List, Tuple


def _header_value(lines: list, name: str) -> str:
    """Return the first value of a header (case-insensitive, compact forms supported)."""
    name_lower = name.lower()
    for line in lines:
        if ':' in line:
            k, _, v = line.partition(':')
            if k.strip().lower() == name_lower:
                return v.strip()
    return ''


def parse(raw: bytes) -> dict:
    """
    Parse raw SIP bytes.

    Returns:
        call_id       – value of the Call-ID header (compact: i)
        request_user  – user part of the INVITE/REFER/ACK Request-URI, empty for responses
        sdp_endpoints – list of (ip, port) tuples from SDP c= / m= lines
        text          – full message decoded as UTF-8 (for writing to trace files)
        first_line    – request-line or status-line
    """
    text = raw.decode('utf-8', errors='replace')

    # Split headers from body (CRLF or LF)
    if '\r\n\r\n' in text:
        header_block, body = text.split('\r\n\r\n', 1)
        header_lines = header_block.split('\r\n')
    elif '\n\n' in text:
        header_block, body = text.split('\n\n', 1)
        header_lines = header_block.split('\n')
    else:
        header_lines = text.split('\n')
        body = ''

    first_line = header_lines[0].strip() if header_lines else ''
    rest = header_lines[1:]

    call_id = _header_value(rest, 'call-id') or _header_value(rest, 'i')

    # Request-URI user — only present in request messages
    request_user = ''
    parts = first_line.split()
    if len(parts) >= 2 and parts[0] not in ('SIP/2.0',):
        # parts[0] is the method, parts[1] is the Request-URI
        m = re.match(r'sips?:([^@>;?\s]+)@', parts[1])
        if m:
            request_user = m.group(1)

    # SDP: walk lines to collect (ip, port) pairs and payload-type → codec mappings
    sdp_endpoints: List[Tuple[str, int]] = []
    rtpmap: dict = {}   # payload_type -> (codec_name, clock_rate)
    current_ip = None
    in_audio = False
    for line in body.replace('\r', '').split('\n'):
        line = line.strip()
        if line.startswith('c='):
            addr = line.split()[-1]
            if addr not in ('0.0.0.0', '::'):
                current_ip = addr
        elif line.startswith('m='):
            in_audio = line.startswith('m=audio')
            if in_audio and current_ip:
                m_parts = line.split()
                if len(m_parts) >= 2:
                    try:
                        port = int(m_parts[1])
                        if port > 0:
                            sdp_endpoints.append((current_ip, port))
                    except ValueError:
                        pass
        elif line.startswith('a=rtpmap:') and in_audio:
            # a=rtpmap:96 opus/48000/2  or  a=rtpmap:0 PCMU/8000
            rest = line[9:]
            pt_str, _, codec_str = rest.partition(' ')
            try:
                pt = int(pt_str)
                parts = codec_str.split('/')
                codec_name = parts[0].upper()
                clock_rate = int(parts[1]) if len(parts) > 1 else 8000
                rtpmap[pt] = (codec_name, clock_rate)
            except (ValueError, IndexError):
                pass

    return {
        'first_line': first_line,
        'call_id': call_id,
        'request_user': request_user,
        'sdp_endpoints': sdp_endpoints,
        'rtpmap': rtpmap,
        'text': text,
    }

File: sipparser.go
package main

import (
	"strconv"
	"strings"
)

// SIPInfo holds parsed information from a SIP message.
type SIPInfo struct {
	CallID       string
	Method       string // request method (INVITE, REGISTER, …); empty for responses
	StatusCode   int    // response status code (100–699); 0 for requests
	StatusReason string // response reason phrase (e.g. "OK", "Busy Here"); empty for requests
	CSeqMethod   string // method from CSeq header (tells which request a response belongs to)
	RequestUser  string
	SDPEndpoints []Endpoint
	RTPMap       map[int]CodecInfo
	Text         string
}

// parseSIP parses raw SIP message bytes and returns a SIPInfo.
// Returns a SIPInfo with empty CallID if the message is not valid SIP.
func parseSIP(payload []byte) *SIPInfo {
	text := string(payload)
	info := &SIPInfo{
		Text:   text,
		RTPMap: make(map[int]CodecInfo),
	}

	// Split headers from body
	var headerBlock, body string
	if idx := strings.Index(text, "\r\n\r\n"); idx >= 0 {
		headerBlock = text[:idx]
		body = text[idx+4:]
	} else if idx := strings.Index(text, "\n\n"); idx >= 0 {
		headerBlock = text[:idx]
		body = text[idx+2:]
	} else {
		headerBlock = text
		body = ""
	}

	// Split header block into lines
	var headerLines []string
	if strings.Contains(headerBlock, "\r\n") {
		headerLines = strings.Split(headerBlock, "\r\n")
	} else {
		headerLines = strings.Split(headerBlock, "\n")
	}

	if len(headerLines) == 0 {
		return info
	}

	firstLine := strings.TrimSpace(headerLines[0])

	// Parse headers (skip first line)
	for _, line := range headerLines[1:] {
		colonIdx := strings.IndexByte(line, ':')
		if colonIdx < 0 {
			continue
		}
		key := strings.TrimSpace(strings.ToLower(line[:colonIdx]))
		val := strings.TrimSpace(line[colonIdx+1:])

		if key == "call-id" || key == "i" {
			if info.CallID == "" {
				info.CallID = val
			}
		} else if key == "cseq" {
			// CSeq: <seq-number> <method>
			if f := strings.Fields(val); len(f) >= 2 {
				info.CSeqMethod = strings.ToUpper(f[1])
			}
		}
	}

	parts := strings.Fields(firstLine)
	if len(parts) >= 2 && parts[0] == "SIP/2.0" {
		// It's a response: "SIP/2.0 CODE Reason phrase..."
		if len(parts) >= 2 {
			if code, err := strconv.Atoi(parts[1]); err == nil {
				info.StatusCode = code
			}
		}
		if len(parts) >= 3 {
			info.StatusReason = strings.Join(parts[2:], " ")
		}
	} else if len(parts) >= 2 {
		// It's a request: parts[0]=method, parts[1]=Request-URI
		info.Method = strings.ToUpper(parts[0])
		uri := parts[1]
		// Match sip:USER@... or sip:USER;... or sip:USER?...
		if sipIdx := strings.Index(strings.ToLower(uri), "sip:"); sipIdx >= 0 {
			userPart := uri[sipIdx+4:]
			// Find end of user (at @, ;, ?, or whitespace)
			end := len(userPart)
			for i, ch := range userPart {
				if ch == '@' || ch == ';' || ch == '?' || ch == ' ' || ch == '\t' {
					end = i
					break
				}
			}
			info.RequestUser = userPart[:end]
		}
	}

	// Parse SDP body
	bodyLines := strings.Split(strings.ReplaceAll(body, "\r", ""), "\n")
	var sdpIP string
	inAudio := false

	for _, line := range bodyLines {
		line = strings.TrimSpace(line)

		if strings.HasPrefix(line, "c=") {
			// c=IN IP4 <IP>
			f := strings.Fields(line)
			if len(f) >= 1 {
				addr := f[len(f)-1]
				if addr != "0.0.0.0" && addr != "::" {
					sdpIP = addr
				}
			}
		} else if strings.HasPrefix(line, "m=") {
			inAudio = strings.HasPrefix(line, "m=audio")
			if inAudio && sdpIP != "" {
				// m=audio <PORT> ...
				f := strings.Fields(line)
				if len(f) >= 2 {
					port, err := strconv.Atoi(f[1])
					if err == nil && port > 0 {
						ep := Endpoint{IP: sdpIP, Port: uint16(port)}
						info.SDPEndpoints = append(info.SDPEndpoints, ep)
					}
				}
			}
		} else if strings.HasPrefix(line, "a=rtpmap:") && inAudio {
			// a=rtpmap:<PT> <NAME>/<RATE>[/<channels>]
			rest := line[9:] // after "a=rtpmap:"
			spaceIdx := strings.IndexByte(rest, ' ')
			if spaceIdx < 0 {
				continue
			}
			ptStr := rest[:spaceIdx]
			codecStr := rest[spaceIdx+1:]

			pt, err := strconv.Atoi(ptStr)
			if err != nil {
				continue
			}

			codecParts := strings.Split(codecStr, "/")
			codecName := strings.ToUpper(codecParts[0])
			clockRate := 8000
			if len(codecParts) > 1 {
				if cr, err := strconv.Atoi(codecParts[1]); err == nil {
					clockRate = cr
				}
			}
			info.RTPMap[pt] = CodecInfo{Name: codecName, ClockRate: clockRate}
		}
	}

	return info
}
